scan-external-contacts/nodejs/index.js

```diff
@@ -0,0 +1,296 @@
+            console.log("Scan External Contacts - cursor: ", currentCursor);
+            let scanResult = await scanContacts(currentCursor);
+
+            if (scanResult.error === false) {
+                if (scanResult.isEmpty === false) {
+                    // Process current data (filter, transform, write)
+                    // Note: filter - You can implement a filter condition here and decide if the entity is to be printed (proceed) or not (return)
+                    // Note: transform - You can manipulate the entity attributes to create the desired string to be printed
+
+                    if (exportType === 'CSV') {
+                        // CSV
+                        if (writeStream && csvStream) {
+                            for (let entity of scanResult.entities) {
+                                // For each entity element, print a line (CSV)
+                                let csvrow = [(entity.id ? entity.id : ''), (entity.modifyDate ? entity.modifyDate : ''), (entity.firstName ? entity.firstName : ''), (entity.lastName ? entity.lastName : ''), (entity.workPhone ? entity.workPhone.e164 : ''), (entity.workEmail ? entity.workEmail : ''), (entity.externalOrganization ? entity.externalOrganization.id : ''), (entity.surveyOptOut ? entity.surveyOptOut.toString() : 'false')];
+                                csvStream.write(csvrow);
+                            }
+                        }
+                    } else {
+                        // RAW JSON
+                        if (writeStream) {
+                            let entitiesString;
+                            if (firstWrite) {
+                                firstWrite = false;
+                                // The entities array is converted to a string, removing first and last character ([ and ])
+                                entitiesString = JSON.stringify(scanResult.entities).slice(1, -1);
+                            } else {
+                                // If some entities have already been printed before, it is necessary to add a "," separator between printed data and new data to print
+                                entitiesString = "," + JSON.stringify(scanResult.entities).slice(1, -1);
+                            }
+
+                            writeStream.write(entitiesString);
+                        }
+                    }
+                }
+
+                let elapsedTime = Date.now() - scanResult.startTime;
+                console.log("Execution of scan - Start At: " + (scanResult.startTime).toString() + " - Duration: " + elapsedTime.toString() + " ms");
+
+                if (scanResult.isFinished === true) {
+                    // Last Data was received (no nextCursor) and export is finished
+                    isCompleted = true;
+                } else {
+                    // Force a pause before next scan (to avoid loading org/token unnecessarily - API rate limit)
+                    let forcedDelay;
```

Comment:
This type of logic is better encapsulated in using a retry code with exponential backoff.  You are trying to show how to use the API not getting buried in the details of pagination logic and error handling :)